worlds/pokemon_emerald/__init__.py

```diff
@@ -0,0 +1,841 @@
+            max_norman_count = 4
+
+        if self.multiworld.hms[self.player].value == RandomizeHms.option_vanilla:
+            if self.multiworld.norman_requirement[self.player].value == NormanRequirement.option_badges:
+                if self.multiworld.badges[self.player].value != RandomizeBadges.option_completely_random:
+                    max_norman_count = 4
+            if self.multiworld.norman_requirement[self.player].value == NormanRequirement.option_gyms:
+                max_norman_count = 4
+
+        self.multiworld.norman_count[self.player].value = min(self.multiworld.norman_count[self.player].value, max_norman_count)
+
+    def create_regions(self) -> None:
+        tags = {"Badge", "HM", "KeyItem", "Rod", "Bike"}
+        if self.multiworld.overworld_items[self.player].value == Toggle.option_true:
+            tags.add("OverworldItem")
+        if self.multiworld.hidden_items[self.player].value == Toggle.option_true:
+            tags.add("HiddenItem")
+        if self.multiworld.npc_gifts[self.player].value == Toggle.option_true:
+            tags.add("NpcGift")
+        if self.multiworld.enable_ferry[self.player].value == Toggle.option_true:
+            tags.add("Ferry")
+
+        create_regions(self.multiworld, self.player)
+        create_locations_with_tags(self.multiworld, self.player, tags)
+
+    def create_items(self) -> None:
+        item_locations: List[PokemonEmeraldLocation] = []
+        for region in self.multiworld.regions:
+            if region.player == self.player:
+                item_locations += region.locations
+
+        # Filter events
+        item_locations = [location for location in item_locations if not location.is_event]
+
+        # Filter progression items which shouldn't be shuffled into the itempool. Their locations
+        # still exist, but event items will be placed and locked at their vanilla locations instead.
+        filter_tags = set()
+
+        if self.multiworld.key_items[self.player].value == Toggle.option_false:
+            filter_tags.add("KeyItem")
+        if self.multiworld.rods[self.player].value == Toggle.option_false:
+            filter_tags.add("Rod")
+        if self.multiworld.bikes[self.player].value == Toggle.option_false:
+            filter_tags.add("Bike")
+
+        if self.multiworld.badges[self.player].value in {RandomizeBadges.option_vanilla, RandomizeBadges.option_shuffle}:
+            filter_tags.add("Badge")
+        if self.multiworld.hms[self.player].value in {RandomizeHms.option_vanilla, RandomizeHms.option_shuffle}:
+            filter_tags.add("HM")
+
+        if self.multiworld.badges[self.player].value == RandomizeBadges.option_shuffle:
+            self.badge_shuffle_info = [
+                (location, self.create_item_by_code(location.default_item_code))
+                for location in [location for location in item_locations if "Badge" in location.tags]
```

Comment:
should definitely use different names for these two separate comprehensions

Reply:
Fixed in 8877f87